### PR TITLE
Recover device

### PR DESCRIPTION
**Pull request by TimeEngineer**

### CODE REVIEW

File: oxidation/libparsec/crates/crypto/src/sodiumoxide/secret.rs [hunk 1]
```diff
@@ -72,13 +66,30 @@ impl SecretKey {
...
         self.0.as_ref()
     }
 }
 
+impl TryFrom<&[u8]> for SecretKey {
+    type Error = CryptoError;
+
+    fn try_from(v: &[u8]) -> Result<Self, Self::Error> {
+        let arr: [u8; Self::SIZE] = v.try_into().map_err(|_| CryptoError::DataSize)?;
+        Ok(Self(Key(arr)))
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/crypto/src/sodiumoxide/secret.rs:83]:
```suggestion
        let arr: [u8; KEYBYTES] = v.try_into().map_err(|_| CryptoError::DataSize)?;
        Ok(Self::from(arr))
```

Re-use the `From<[u8; KEYBYTES]>`

File: oxidation/libparsec/crates/crypto/src/macros.rs [hunk 2]
```diff
@@ -1,5 +1,9 @@
 // Parsec Cloud (https://parsec.cloud) Copyright (c) BUSL-1.1 (eventually AGPL-3.0) 2016-present Scille SAS
 
+#[cfg(any(
+    all(feature = "use-rustcrypto", not(feature = "use-sodiumoxide")),
+    all(feature = "use-sodiumoxide", not(feature = "use-rustcrypto")),
+))]
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/crypto/src/macros.rs:6]: Yes but warning can happen

File: oxidation/libparsec/crates/crypto/src/macros.rs [hunk 3]
```diff
@@ -11,4 +15,8 @@ macro_rules! impl_key_debug {
     };
 }
 
+#[cfg(any(
+    all(feature = "use-rustcrypto", not(feature = "use-sodiumoxide")),
+    all(feature = "use-sodiumoxide", not(feature = "use-rustcrypto")),
+))]
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/crypto/src/macros.rs:21]: duplicated

File: oxidation/libparsec/crates/crypto/src/lib.rs [hunk 4]
```diff
@@ -1,10 +1,30 @@
...
+#[cfg(all(feature = "use-sodiumoxide", feature = "use-rustcrypto"))]
+compile_error!("features `use-sodiumoxide` and `use-rustcrypto` are mutually exclusive");
+
+mod common;
 mod error;
-#[cfg_attr(feature = "use-sodiumoxide", path = "sodiumoxide/mod.rs")]
-#[cfg_attr(feature = "use-rustcrypto", path = "rustcrypto/mod.rs")]
-mod implementation;
 mod macros;
+#[cfg(all(feature = "use-rustcrypto", not(feature = "use-sodiumoxide")))]
```
[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/crypto/src/lib.rs:11]: Errors can happen then hide the first error message

File: oxidation/libparsec/crates/crypto/src/sodiumoxide/secret.rs (hunk 1 above)
Comment (TimeEngineer) [reply oxidation/libparsec/crates/crypto/src/sodiumoxide/secret.rs]: Done

File: oxidation/libparsec/crates/crypto/src/lib.rs (hunk 4 above)
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/crypto/src/lib.rs:11]: Rustc won't stop at the compile_error ?

**Review by Volham22 — APPROVED** [review]

File: oxidation/libparsec/crates/crypto/src/lib.rs [hunk 5]
```diff
@@ -1,10 +1,30 @@
...
+mod rustcrypto;
+#[cfg(all(feature = "use-sodiumoxide", not(feature = "use-rustcrypto")))]
+mod sodiumoxide;
```
[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/crypto/src/lib.rs:14]: How should I consider this reply ?

[the remark this answers is not in the document]
Comment (TimeEngineer) [reply oxidation/libparsec/crates/crypto/src/lib.rs:14]: Did you try ?

[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/crypto/src/lib.rs:14]: Already ;)

https://github.com/Scille/parsec-cloud/blob/b5abc38e5cd92214dae9200b5d4c7f87f2ca17a3/oxidation/libparsec/crates/platform_async/src/lib.rs#L3-L11

[the remark this answers is not in the document]
Comment (FirelightFlagboy) [reply oxidation/libparsec/crates/crypto/src/lib.rs:14]: But yes it's not in the `mod` but in the `use`

File: oxidation/libparsec/crates/crypto/src/sodiumoxide/secret.rs [hunk 6]
```diff
@@ -72,13 +66,30 @@ impl SecretKey {
...
     }
 }
 
+impl TryFrom<&[u8]> for SecretKey {
+    type Error = CryptoError;
+
+    fn try_from(v: &[u8]) -> Result<Self, Self::Error> {
+        let arr: [u8; KEYBYTES] = v.try_into().map_err(|_| CryptoError::DataSize)?;
+        Ok(Self(Key(arr)))
+    }
```
Comment (FirelightFlagboy) [inline oxidation/libparsec/crates/crypto/src/sodiumoxide/secret.rs:84]: Use `From<[u8; KEYBYTES]>`

**Pull request merged**